### DISCUSSION TITLE
Exclude assembly that is added by consumed package targets

### DISCUSSION THREAD

**Discussion by quasarea** [devlooped/nugetizer#120] [Q&A] [answered]
Hi, 

I have library that depends on Grpc.Core package. This package has a targets file that adds assemblies to output 
```
...
  <ItemGroup Condition="'$(Grpc_SkipNativeLibsCopy)' != 'true'">
    <Content Include="$(MSBuildThisFileDirectory)..\..\runtimes\win-x86\native\grpc_csharp_ext.x86.dll">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <Link>grpc_csharp_ext.x86.dll</Link>
      <Visible>false</Visible>
    </Content>
...
```

How can I exclude grpc_csharp_ext.x86.dll from being packaged? I have tried several combinations of 
```
    <PackInference Update="Content" PackExclude="**/grpc_csharp_ext.x86.dll;**/grpc_csharp_ext.x64.dll" />
    <PackInference Update="Content" PackExclude="$(OutputPath)/**/grpc_csharp_ext.x64.dll" />
    <PackInference Update="Content" PackExclude="$(OutputPath)/grpc_csharp_ext.x64.dll" />
    <PackInference Update="Content" PackExclude="**/grpc_csharp_ext.x64.dll" />
```
But nothing worked :(

**Comment by kzu** [comment]
Ok, given this project:

```xml
<Project Sdk="Microsoft.NET.Sdk">

	<PropertyGroup>
		<TargetFramework>netstandard2.1</TargetFramework>
		<PackageId>GrpcLib</PackageId>
	</PropertyGroup>

	<ItemGroup>
		<PackageReference Include="Grpc.Core" Version="2.38.1" />
		<PackageReference Include="NuGetizer" Version="0.7.2" />
	</ItemGroup>

</Project>
```

This is what `nugetize` shows:

![image](https://user-images.githubusercontent.com/169707/126006381-23ae159a-ee15-436c-878c-421c7acf7b06.png)

which matches what's packed, so I'd say it's working as expected. could you perhaps provide a repro?

**Reply by quasarea** [reply]
Sorry, should mention that, needs to be .net framework 4.5+ to repro, as targets in package are:
![image](https://user-images.githubusercontent.com/10334982/126022438-b7d22785-d812-4258-8d50-f1f5bdebdd33.png)

**Reply by quasarea** [reply]
Project attached, but its just blank 4.8 with dependencies as above.
[NugetizerIgnore.zip](https://github.com/devlooped/nugetizer/files/6840598/NugetizerIgnore.zip)

```
PS C:\source\NugetizerIgnore> nugetize
Project C:\source\NugetizerIgnore\NugetizerIgnore.sln is not packable, rendering its contributed package contents.
  Dependencies:
    net472
      Grpc.Core, 2.36.4
  Contents:
    /lib/
      net472/
        grpc_csharp_ext.x64.dll
        grpc_csharp_ext.x86.dll
        libgrpc_csharp_ext.x64.dylib
        libgrpc_csharp_ext.x64.so
        NugetizerIgnore.exe
        NugetizerIgnore.exe.config
        NugetizerIgnore.pdb
```

**Reply by kzu** [reply]
Ok, to make the "fix" reliable (note: this is not a nugetizer issue, since arbitrary targets can add arbitrary files), I submitted this PR to Grps.Core: https://github.com/grpc/grpc/pull/26725.

The solution you are looking for (given the currently shipping Grpc.Core) would be:

```xml
	<Target Name="ExcludeGrpcContentFromPack" BeforeTargets="InferPackageContents">
		<ItemGroup>
			<Content Update="@(Content -> WithMetadataValue('DefiningProjectName', 'Grpc.Core'))" Pack="false" />
		</ItemGroup>
	</Target>
```

When/if they accept my PR, you'd change `DefiningProjectName` with `NuGetPackageId` and that's it. 

Here's the output from `nugetize` with that target in place:

![image](https://user-images.githubusercontent.com/169707/126209972-057c00f5-a58d-41ad-affd-993e6edb1498.png)

**Reply by kzu** [reply]
I'm currently investigating why the update to PackInference you made didn't kick in, also.

**Comment by kzu** [comment, marked as answer]
The [related bug](https://github.com/devlooped/nugetizer/pull/129) for this has shipped and allows you to exclude all these library items with:

```xml
<PackInference Update="Content" PackExclude="**/*grpc_csharp_ext*.*" />
```

**Reply by quasarea** [reply]
Would you expect it to work on everything, like if I would like to not package pdb from the build, would the below work?
```
<ItemGroup>
   <PackInference Update="Content" PackExclude="**/*.pdb" />
</ItemGroup>
```
Asking as it doesn't in 4.8, unless I'm doing something totally wrong here.

I want to generate pdb, for in example binskim, but in some cases I don't want to expose it.

**Reply by kzu** [reply]
Pdbs are part of the primary build output, not content. For that scenario you would just set PackSymbols=false instead as documented in the readme

**Comment by quasarea** [comment]
Hi, sorry, I'm still struggling with exclusion (grpc included your change in current prerelease package.

I added custom target to update it like you demonstrated:
```
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003" DefaultTargets="Build" ToolsVersion="15.0">
  <Target Name="ExcludeGrpcContentFromPack" BeforeTargets="InferPackageContents">
    <Message Importance="high" Text="--- ExcludeGrpcContentFromPack ---" />
    <Message Importance="high" Text="@(Content -> WithMetadataValue('NuGetPackageId', 'Grpc.Core'))" />
    <ItemGroup>
      <Content Update="@(Content -> WithMetadataValue('NuGetPackageId', 'Grpc.Core'))" Pack="false" />
    </ItemGroup>
  </Target>
</Project>
```

Target is executed during both nugetize and msbuild 
```
PS C:\git\ExclusionTest> nugetize
  Determining projects to restore...
  Restored C:\git\ExclusionTest\ExclusionTest\ExclusionTest.csproj (in 152 ms).
  --- ExcludeGrpcContentFromPack ---
  C:\Users\user\.nuget\packages\grpc.core\2.41.0-pre2\build\net45\..\..\runtimes\win-x86\native\grpc_csharp_ext.x86.dll;
  C:\Users\user\.nuget\packages\grpc.core\2.41.0-pre2\build\net45\..\..\runtimes\win-x64\native\grpc_csharp_ext.x64.dll;
  C:\Users\user\.nuget\packages\grpc.core\2.41.0-pre2\build\net45\..\..\runtimes\linux-x64\native\libgrpc_csharp_ext.x64.so;
  C:\Users\user\.nuget\packages\grpc.core\2.41.0-pre2\build\net45\..\..\runtimes\osx-x64\native\libgrpc_csharp_ext.x64.dylib
Package: ExclusionTest.0.0.1.nupkg
         C:\git\ExclusionTest\ExclusionTest\bin\ExclusionTest.nuspec
    Authors    : Authors
    Description: Package Description
    Version    : 0.0.1
  Dependencies:
    net48
      Grpc.Core, 2.41.0-pre2
  Contents:
    /lib/
      net48/
        ExclusionTest.dll
        ExclusionTest.pdb
        grpc_csharp_ext.x64.dll
        grpc_csharp_ext.x86.dll
        libgrpc_csharp_ext.x64.dylib
        libgrpc_csharp_ext.x64.so
```

Here is my sample project with targets and rsp so just do `msbuild`

[ExclusionTest.zip](https://github.com/devlooped/nugetizer/files/7246022/ExclusionTest.zip)

**Reply by quasarea** [reply]
@kzu could it be because we are excluding assemblies from nuget global folder, while nugetizer (in generated nuspec) adds same assemblies but from local build folder
```
...
    <file src="C:\git\ExclusionTest\Build\grpc_csharp_ext.x86.dll" target="lib\net48\grpc_csharp_ext.x86.dll" />
    <file src="C:\git\ExclusionTest\Build\grpc_csharp_ext.x64.dll" target="lib\net48\grpc_csharp_ext.x64.dll" />
    <file src="C:\git\ExclusionTest\Build\libgrpc_csharp_ext.x64.so" target="lib\net48\libgrpc_csharp_ext.x64.so" />
    <file src="C:\git\ExclusionTest\Build\libgrpc_csharp_ext.x64.dylib" target="lib\net48\libgrpc_csharp_ext.x64.dylib" />
```

**Comment by quasarea** [comment]
Ok, the 
``` 
  <ItemGroup>
    <PackInference Update="Content" PackExclude="**/*grpc_csharp_ext.*" />
  </ItemGroup>
```
works fine now, so I will use this instead of target based exclusion

**Reply by kzu** [reply]
I sent this PR upstream too and got merged, so that can also provide an alternative: https://github.com/grpc/grpc/pull/26725 (using the PackageId metadata for those files)
